### PR TITLE
[processing][feature] Add a row_index and col_index field to the output of the "Create Grid" algorithm

### PR DESCRIPTION
**Pull request by ValentinBuira**
## Description

This pull request add a row_index and col_index as field to the output of the "Create Grid" algorithm, for the output of type point, rectangle and hexagon

preview :
screenshot_grid_qgis

row and col variable where already used for generating the grid, this commit just expose them into new fields

Motivation for this change: In my use case I had a grid with data associated to each polygons and I needed to calculate the moving average to smooth de data.

### CODE REVIEW

**Comment by Poulpator** [comment]
Hi,

Would it be possible to launch the workflows awaiting for approval, with the aim to merge this pull request at the end

poke @nyalldawson since you look active in pull requests review, but anyone feel free to do it

File: src/analysis/processing/qgsalgorithmgrid.cpp [hunk 1]
```diff
@@ -119,6 +119,18 @@ QVariantMap QgsGridAlgorithm::processAlgorithm( const QVariantMap &parameters, Q
...
   fields.append( QgsField( QStringLiteral( "bottom" ), QVariant::Double ) );
+  
+  switch (mIdx)
+  {
+  case 0: //point
+  case 2: //rectangle
+  case 4: //hexagon
+    fields.append( QgsField( QStringLiteral( "row_index" ), QVariant::LongLong ) );
+    fields.append( QgsField( QStringLiteral( "col_index" ), QVariant::LongLong ) );
+    break;
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmgrid.cpp:130]:
```suggestion
    case 0: //point
    case 2: //rectangle
    case 4: //hexagon
      fields.append( QgsField( QStringLiteral( "row_index" ), QVariant::LongLong ) );
      fields.append( QgsField( QStringLiteral( "col_index" ), QVariant::LongLong ) );
      break;
    default:
      break;  
```

File: src/analysis/processing/qgsalgorithmgrid.cpp [hunk 2]
```diff
@@ -186,7 +198,7 @@ void QgsGridAlgorithm::createPointGrid( std::unique_ptr< QgsFeatureSink > &sink,
       const double y = mGridExtent.yMaximum() - ( row * mVSpacing - row * mVOverlay );
 
       f.setGeometry( QgsGeometry( new QgsPoint( x, y ) ) );
-      f.setAttributes( QgsAttributes() << id << x << y << x + mHSpacing << y + mVSpacing );
+      f.setAttributes( QgsAttributes() << id << x << y << x + mHSpacing << y + mVSpacing << row << col);
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmgrid.cpp:201]:
```suggestion
      f.setAttributes( QgsAttributes() << id << x << y << x + mHSpacing << y + mVSpacing << row << col );
```

File: src/analysis/processing/qgsalgorithmgrid.cpp [hunk 3]
```diff
@@ -346,7 +358,7 @@ void QgsGridAlgorithm::createRectangleGrid( std::unique_ptr< QgsFeatureSink > &s
       std::unique_ptr< QgsPolygon > poly = std::make_unique< QgsPolygon >();
       poly->setExteriorRing( new QgsLineString( ringX, ringY ) );
       f.setGeometry( std::move( poly ) );
-      f.setAttributes( QgsAttributes() << id << x1 << y1 << x2 << y2 );
+      f.setAttributes( QgsAttributes() << id << x1 << y1 << x2 << y2 << row << col);
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmgrid.cpp:361]:
```suggestion
      f.setAttributes( QgsAttributes() << id << x1 << y1 << x2 << y2 << row << col );
```

File: src/analysis/processing/qgsalgorithmgrid.cpp [hunk 4]
```diff
@@ -119,6 +119,20 @@ QVariantMap QgsGridAlgorithm::processAlgorithm( const QVariantMap &parameters, Q
   fields.append( QgsField( QStringLiteral( "top" ), QVariant::Double ) );
   fields.append( QgsField( QStringLiteral( "right" ), QVariant::Double ) );
   fields.append( QgsField( QStringLiteral( "bottom" ), QVariant::Double ) );
+  
+  switch (mIdx)
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmgrid.cpp:123]:
```suggestion
  switch ( mIdx )
```

File: src/analysis/processing/qgsalgorithmgrid.cpp [hunk 5]
```diff
@@ -119,6 +119,20 @@ QVariantMap QgsGridAlgorithm::processAlgorithm( const QVariantMap &parameters, Q
...
+  {
+    case 0: //point
+    case 2: //rectangle
+    case 4: //hexagon
+      fields.append( QgsField( QStringLiteral( "row_index" ), QVariant::LongLong ) );
+      fields.append( QgsField( QStringLiteral( "col_index" ), QVariant::LongLong ) );
+      break;
+    default:
+      break;
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmgrid.cpp:132]:
```suggestion
      break;
```

File: src/analysis/processing/qgsalgorithmgrid.cpp [hunk 6]
```diff
@@ -119,6 +119,20 @@ QVariantMap QgsGridAlgorithm::processAlgorithm( const QVariantMap &parameters, Q
...
+  }
+  
+
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmgrid.cpp:135]:
```suggestion

  switch ( mIdx )
  {
    case 0: //point
    case 2: //rectangle
    case 4: //hexagon
      fields.append( QgsField( QStringLiteral( "row_index" ), QVariant::LongLong ) );
      fields.append( QgsField( QStringLiteral( "col_index" ), QVariant::LongLong ) );
      break;
    default:
      break;
  }

```

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**